# Time-to-event TMLE: plan exit, readmission, and death under repeated navigation

This notebook estimates the risk of an event at two horizons with longitudinal targeted maximum
likelihood estimation (TMLE). Each step shows its code, its output, and what the output tells you.
The [technical entry](../technical-reference/longitudinal-tmle.md#survival-and-competing-risks)
gives the event-process recursion.

## The applied question

A patient who leaves the plan is not a patient with a missing outcome. The plan's retention team
asks two questions about repeated navigation.

| question | the outcome | the step that answers it |
| --- | --- | --- |
| does navigation in each period lower the risk of plan exit at 30 and 60 days? | plan exit for any reason, including death | Steps 6 and 7 |
| how does navigation change readmission, when death can end the time at risk first? | readmission and death as competing events, with disenrollment as censoring | Steps 8 and 9 |

The first outcome counts death as plan exit. A lower exit risk can therefore come from fewer deaths
alone, because this outcome does not separate death from disenrollment. The second question
separates readmission from death.

Time zero is hospital discharge, and each period is 30 days. The program offers navigation at the
start of each period. The second decision is therefore on day 31, not on day seven. This page
reuses the design vocabulary of [longitudinal TMLE](longitudinal-tmle.ipynb), which introduces the
sequential regression.

## What you will learn

| after this notebook you can | the step that shows it |
| --- | --- |
| say why a comparison of plan followers is not a causal effect | Step 3 |
| write a protocol for an event outcome | Step 4 |
| declare an absorbing event, loss of tracking, and navigator teams | Step 5 |
| read one cumulative risk per plan per horizon | Step 6 |
| compare two plans from one fit, and read the survival view | Step 7 |
| declare two competing events and disenrollment as censoring, and read each cause's incidence | Steps 8 and 9 |
| see why death coded as censoring answers a different question | Step 10 |
| read what the assessment reports for a longitudinal fit | Step 11 |
| measure how far the estimate moves without each recorded covariate | Step 12 |

## Why this method

After a patient exits, the later navigation offer and needs measurement do not exist. A plan can
therefore intervene only on patients who are still at risk. Each horizon is its own parameter, and
each node's regression uses the patients at risk entering that node.

| your situation | what this method buys | what it costs |
| --- | --- | --- |
| the outcome is an event that can happen at more than one time | one cumulative risk per horizon, each on its own risk set | one regression per node per horizon |
| patients can experience one of two competing events | a cause-specific cumulative incidence, with the competing cause left in the history | one backward pass per cause. The estimated incidences are not constrained to sum to the all-cause risk |
| you want the retention scale rather than the risk scale | `curve(scale="survival")`, which mirrors each level and its interval | nothing. It is the same fit read the other way |

The table below defines the terms this notebook uses most. Each step links to the full definition
where the term first matters.

| term | plain meaning |
| --- | --- |
| node | one time in the sequence at which a column is recorded, such as the day-31 offer |
| cumulative risk | the probability that the event has happened by a horizon, under a plan |
| absorbing event | an event after which the patient stays in that state, such as plan exit |
| risk set | the patients who have not had the event and are still tracked when a node starts |
| censoring | the plan loses track of a patient before it can see whether the event happened |
| competing event | an event that makes the event of interest impossible afterwards, such as death before readmission |
| cause-specific cumulative incidence | the probability that one cause has ended the time at risk by a horizon |
| influence curve | how much each row moves the estimate. Its variance gives the standard error |

## Step 1: set up

The setup imports the learners and prints the installed `cleverly` version. Every fit below passes
its learners and random seed explicitly, so a rerun reproduces the stored outputs.

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression, LogisticRegression

import cleverly

pd.set_option("display.width", 160)
pd.set_option("display.max_columns", 20)
print("cleverly", cleverly.__version__)

cleverly 0.1.2


**What this output tells you.** The stored outputs in this notebook came from the version named
above. A different version can print different numbers.

## Step 2: the data

The data come from a synthetic law with a known answer. The code renames the generator's columns
to the program's names. It prints the first rows, counts the patients who leave each risk set, and
prints the true risks of the law.

In [2]:
from cleverly.datasets import make_longitudinal_survival

exit_frame, exit_truth = make_longitudinal_survival(n=4_000, seed=52, cluster_size=20)
exit_frame = exit_frame.rename(
    columns={
        "W1": "age",
        "W2": "baseline_readiness",
        "A1": "navigation_p1",
        "C1": "tracked_p1",
        "Y1": "plan_exit_p1",
        "L2": "identified_needs",
        "A2": "navigation_p2",
        "C2": "tracked_p2",
        "Y2": "plan_exit_p2",
        "id": "navigator_team",
    }
)
print("rows and columns:", exit_frame.shape)
print(exit_frame.head().round(3).to_string())
print()
print("navigator teams:              ", exit_frame["navigator_team"].nunique())
print("lost to tracking in period 1: ", int(exit_frame["tracked_p1"].eq(0).sum()))
print("exited in period 1:           ", int(exit_frame["plan_exit_p1"].eq(1).sum()))
print("at risk entering period 2:    ", int(exit_frame["plan_exit_p1"].eq(0).sum()))
print("lost to tracking in period 2: ", int(exit_frame["tracked_p2"].eq(0).sum()))
print()
print("known values of the synthetic law, cumulative risk of plan exit:")
for key in (
    "risk_regimen[never @ t=1]",
    "risk_regimen[never @ t=2]",
    "risk_regimen[always @ t=1]",
    "risk_regimen[always @ t=2]",
    "ate_regimen[always vs never @ t=1]",
    "ate_regimen[always vs never @ t=2]",
):
    print(f"  {key:36s} {exit_truth[key]:.4f}")

rows and columns: (4000, 10)
     age  baseline_readiness  navigation_p1  tracked_p1  plan_exit_p1  identified_needs  navigation_p2  tracked_p2  plan_exit_p2  navigator_team
0 -0.826               0.743            0.0         1.0           1.0               NaN            NaN         NaN           1.0             0.0
1  0.243               1.038            0.0         1.0           1.0               NaN            NaN         NaN           1.0             0.0
2 -0.064              -0.118            0.0         1.0           0.0             1.003            0.0         1.0           0.0             0.0
3 -0.094               0.063            1.0         1.0           0.0             2.693            1.0         1.0           0.0             0.0
4  0.323              -1.698            0.0         1.0           1.0               NaN            NaN         NaN           1.0             0.0

navigator teams:               200
lost to tracking in period 1:  463
exited in period 1:           

**What this output tells you.** Each row is one patient, and 20 patients share each of the 200
navigator teams. The generator is `make_longitudinal_survival`. The columns are in time order.

| column | node | role |
| --- | --- | --- |
| `age`, `baseline_readiness` | before discharge | standardized baseline covariates (mean 0, SD 1) |
| `navigation_p1`, `navigation_p2` | start of each period | the navigation offer |
| `tracked_p1`, `tracked_p2` | during each period | 1 if the plan can observe exit in that period |
| `plan_exit_p1`, `plan_exit_p2` | end of each period | the absorbing event, carried forward after exit |
| `identified_needs` | end of period 1, among patients at risk | raised by navigation, raises day-31 navigation and the exit hazard |
| `navigator_team` | fixed | the cluster |

The plan learns of an exit from the monthly eligibility files that employer groups send. When a
group stops sending its file, the plan cannot see whether those members left. Such a patient is
lost to tracking.

Rows 0, 1, and 4 exited in period 1. Their period-2 nodes are `NaN` because those nodes do not
exist, and `plan_exit_p2` carries the 1 forward. The counts show the two ways to leave the first
risk set: 463 patients were lost to tracking and 739 exited. That leaves 2798 patients at risk
entering period 2.

In this synthetic law, 26% of patients would exit within 30 days and 46% within 60 days under no
navigation. The true risk differences of always against never are -0.1082 and -0.1380. The law
sets high exit rates, so each fit has many events to model. A real program has no `truth`, and
every comparison against it below is a teaching device.

## Step 3: association first

A confounder changes both who receives navigation and the outcome. Here confounding happens at
both decisions. The code compares the groups at each decision. It then compares the observed exit
share of patients whose recorded offers followed each plan.

In [3]:
print("baseline means by the discharge offer:")
print(exit_frame.groupby("navigation_p1")[["age", "baseline_readiness"]].mean().round(3))
print()
at_risk_p2 = exit_frame["plan_exit_p1"].eq(0)
needs = exit_frame.loc[at_risk_p2].groupby("navigation_p2")["identified_needs"].mean()
print("mean identified_needs by the day-31 offer, among patients at risk:")
print(needs.round(3))
print()

# A follower's recorded offers match the plan in every period the patient was at risk.
first = {
    "always": exit_frame["navigation_p1"].eq(1),
    "never": exit_frame["navigation_p1"].eq(0),
}
second = {
    "always": ~exit_frame["navigation_p2"].eq(0),
    "never": ~exit_frame["navigation_p2"].eq(1),
}
followers = {1: first, 2: {plan: first[plan] & second[plan] for plan in first}}
naive_differences = {}
print("observed exit share among tracked plan followers:")
for horizon in (1, 2):
    outcome = exit_frame[f"plan_exit_p{horizon}"]
    shares = {
        plan: float(outcome[rows & outcome.notna()].mean())
        for plan, rows in followers[horizon].items()
    }
    naive_differences[horizon] = shares["always"] - shares["never"]
    population = exit_truth[f"ate_regimen[always vs never @ t={horizon}]"]
    print(
        f"  t={horizon}: always {shares['always']:.4f}, never {shares['never']:.4f}, "
        f"difference {naive_differences[horizon]:.4f}, population {population:.4f}"
    )

baseline means by the discharge offer:
                 age  baseline_readiness
navigation_p1                           
0.0           -0.143               0.202
1.0            0.151              -0.206

mean identified_needs by the day-31 offer, among patients at risk:
navigation_p2
0.0   -0.059
1.0    0.825
Name: identified_needs, dtype: float64

observed exit share among tracked plan followers:
  t=1: always 0.1696, never 0.2489, difference -0.0793, population -0.1082
  t=2: always 0.4232, never 0.5483, difference -0.1251, population -0.1380


**What this output tells you.** The groups differ before each offer.

| comparison | offered | not offered | what it means |
| --- | --- | --- | --- |
| mean `age` at the discharge offer | 0.151 | -0.143 | offered patients are older |
| mean `baseline_readiness` at the discharge offer | -0.206 | 0.202 | offered patients are less ready |
| mean `identified_needs` at the day-31 offer | 0.825 | -0.059 | the day-31 offer goes to patients with more needs |

In this law, older age, lower readiness, and more identified needs all raise the exit hazard. The
offered patients therefore start at higher risk. The follower comparison gives -0.0793 at 30 days
and -0.1251 at 60 days. The true differences are -0.1082 and -0.1380, so the naive comparison
understates the benefit on this draw. It also drops every patient the plan lost track of.

`identified_needs` is a time-varying confounder. Navigation raises it, and it drives the next
offer. [Longitudinal TMLE](longitudinal-tmle.ipynb#step-7-the-failure-mode-a-point-treatment-analysis-of-the-same-data)
shows why one regression cannot adjust for it. The next steps state the question, and the
assumptions that make a causal comparison possible.

## Step 4: write the protocol

A `StudyProtocol` records the scientific design before any model runs.
[Point-treatment TMLE](point-treatment-tmle.ipynb#step-4-write-the-protocol) introduces its
fields and its fingerprint. The code starts from `longitudinal_navigation_protocol()`, the
two-decision program of the [shared study design](index.md#the-shared-study-design). It then
replaces only the fields that an event outcome with 30-day periods changes.

In [4]:
from dataclasses import replace

from cleverly.datasets import longitudinal_navigation_protocol

program = longitudinal_navigation_protocol()
exit_protocol = replace(
    program,
    eligibility=(*program.eligibility, "Enrolled in the health plan at discharge"),
    treatment_strategies=(
        "Offer navigation at the start of each 30-day period while enrolled",
        program.treatment_strategies[1],
    ),
    treatment_versions=(
        "The declared navigation contacts at discharge and on day 31",
        program.treatment_versions[1],
    ),
    outcome="Plan exit for any reason, recorded at the end of each 30-day period",
    horizon="30 and 60 days after discharge",
    intercurrent_event_handling=(
        "Count death as plan exit (composite strategy)",
        "Make no navigation decision after plan exit",
        program.intercurrent_event_handling[2],
    ),
    assumption_rationale=(
        "Recorded history covers the measured common causes of navigation, loss of "
        "tracking, and plan exit in each period",
        *program.assumption_rationale[1:],
    ),
)
print("\n".join(exit_protocol.summary_lines()))

causal study protocol: schema 1; 67439090fc4bcdb0
target population: Adults discharged home from a participating hospital during the enrollment period
eligibility: ['Age 18 years or older', 'Discharged alive', 'Discharged home from a participating hospital', 'Enrolled in the health plan at discharge']
time zero: Hospital discharge, after baseline measurement and before first assignment
treatment strategies: ['Offer navigation at the start of each 30-day period while enrolled', 'Offer no navigation at either decision']
treatment versions: ['The declared navigation contacts at discharge and on day 31', 'Usual discharge support without navigation contacts']
outcome: Plan exit for any reason, recorded at the end of each 30-day period
horizon: 30 and 60 days after discharge
intercurrent-event handling: ['Count death as plan exit (composite strategy)', 'Make no navigation decision after plan exit', 'Analyze each navigation offer regardless of completed contacts']
interference unit: Individua

**What this output tells you.** The first line gives the schema version and the fingerprint
`67439090fc4bcdb0`. The other lines repeat each field. The code changes seven fields of the
program protocol and keeps three.

| protocol field | what this page writes, and why |
| --- | --- |
| target population, time zero, interference unit | unchanged. Discharge aligns eligibility, the first offer, and follow-up |
| eligibility | adds enrollment in the health plan at discharge, because only an enrolled patient can exit |
| treatment strategies and versions | "always" offers navigation at the start of each 30-day period while the patient stays enrolled. The second offer is on day 31 |
| outcome and horizon | plan exit at the end of each 30-day period, read at 30 and 60 days |
| intercurrent-event handling | death counts as plan exit, and no decision exists after exit |
| assumption rationale | the first entry names loss of tracking and plan exit in each period |

Three design choices have no protocol field. The `censoring=` role in the next step declares loss
of tracking. The `cluster=` role declares the navigator teams, which are dependence, not
interference. The typed estimand declares the plans and the horizons.

## Step 5: design and identification

The design places every column at its node. The `outcome=` role carries the event. One outcome
column per time point declares an absorbing event. The `censoring=` role carries whether the plan
could observe the patient in that period. A patient the plan lost track of is censored. A patient
who left the plan had the event.

The code keeps the treatment and history nodes in `navigation_nodes`, because every design on this
page reuses them. `RegimeMean` asks for the cumulative risk under each plan. The
[estimands guide](../user-guide/estimands.md) defines an estimand: the number the question asks
for, written before any model is chosen.

In [5]:
from cleverly import CausalStudy, LongitudinalTreatment, RegimeMean

# Every design on this page declares the same navigation offers and the same history.
navigation_nodes = {
    "treatment": ("navigation_p1", "navigation_p2"),
    "baseline": ("age", "baseline_readiness"),
    "time_varying": ((), ("identified_needs",)),
}
plans = {"always": 1, "never": 0}
levels_by_plan = RegimeMean(plans, reference="never", horizons=(1, 2))
exit_study = CausalStudy(
    exit_frame,
    design=LongitudinalTreatment(
        outcome=("plan_exit_p1", "plan_exit_p2"),
        censoring=("tracked_p1", "tracked_p2"),
        cluster="navigator_team",
        **navigation_nodes,
    ),
    protocol=exit_protocol,
)
exit_effect = exit_study.identify(levels_by_plan)
print(exit_effect.summary(protocol="fingerprint"))

mean outcome under each declared regime
identified by explicit-adjustment: sequential g-formula under the declared treatment regimen
history at navigation_p1: ['age', 'baseline_readiness']
history at navigation_p2: ['age', 'baseline_readiness', 'identified_needs', 'navigation_p1']
required nuisances: ['sequential_outcome_regressions', 'treatment_and_censoring_mechanisms']
assumptions:
  - consistency: each observed history equals the potential history under its realized regimen
  - no interference: one unit's potential history does not depend on other units' regimens
  - sequential exchangeability given the recorded history at every node
  - sequential positivity for treatment and remaining under observation
causal study protocol: schema 1; 67439090fc4bcdb0


**What this output tells you.** The first lines name the estimand and its identification, the
sequential g-formula. The summary then prints the fingerprint of the stored protocol.

The two `history at` lines give the history at each navigation node. Each line lists the
covariates first and the earlier offers last. The second node adds `identified_needs`, which
`time_varying=` places before that node, and the first offer.

The required nuisances are the sequential outcome regressions and the treatment-and-censoring
mechanisms. A nuisance is a model the estimate needs but the question does not ask about.
[Point-treatment TMLE](../technical-reference/point-treatment-tmle.md) introduces the term.

The four assumptions change shape for an event outcome.

| assumption | what it becomes here | can the data check it? |
| --- | --- | --- |
| exchangeability | sequential. At every node, navigation and remaining tracked are independent of the potential outcomes, given the recorded history | no |
| positivity | sequential, and per horizon. A later horizon passes through every earlier node, so its cumulative product has more factors | partly, through the support report |
| consistency | each period uses the declared protocol version, and the recorded event is the event under that protocol | no |
| no interference | one patient's assignments do not change another patient's protocol or outcome | no |

Positivity means that every kind of patient has some chance of following each plan.
[Diagnostics](../user-guide/results-assessment.md#diagnostics) describes what the support report
can show about it.

## Step 6: estimate the cumulative risks

Every fit on this page uses the same parametric learners, fitted on all rows without
cross-fitting.

| setting | value | what it does |
| --- | --- | --- |
| `outcome_learner` | logistic regression | fits the regression whose outcome is the recorded event |
| `pseudo_learner` | linear regression | fits the intermediate regressions, whose outcome is a bounded prediction |
| `treatment_learner` | logistic regression | fits the probability of each navigation offer |
| `censoring_learner` | logistic regression | fits the probability of remaining tracked |
| `CrossFitting(enabled=False)` | in-sample | fits each nuisance on all rows |
| `Runtime(random_state=41, n_jobs=1)` | fixed seed, one process | makes the fit reproducible |

Parametric GLM learners satisfy the Donsker condition. That condition limits how flexible a
learner can be when it predicts the rows it was fitted on. The in-sample fits therefore keep
influence-curve inference valid under the
[remaining conditions](../technical-reference/cv-tmle.md#what-this-solves). Cross-fitting, where
each row's prediction comes from models fit without that row, is the alternative for flexible
learners.

Targeting then updates each regression. The clever covariate uses the inverse of the cumulative
product of the probabilities of following the plan and staying tracked through that node.
[The longitudinal algorithm](../technical-reference/longitudinal-tmle.md#the-algorithm-as-implemented)
defines that covariate; [targeting and bounds](../user-guide/methods-learners.md#targeting-and-bounds)
gives the configuration options.

In [6]:
from cleverly import CrossFitting, ModelSpec, Runtime, TMLEMethod

sequential = TMLEMethod(
    models=ModelSpec(
        outcome_learner=LogisticRegression(max_iter=1000, random_state=41),
        pseudo_learner=LinearRegression(),
        treatment_learner=LogisticRegression(max_iter=1000, random_state=41),
        censoring_learner=LogisticRegression(max_iter=1000, random_state=41),
    ),
    cross_fitting=CrossFitting(enabled=False),
    runtime=Runtime(random_state=41, n_jobs=1),
)
exit_result = exit_effect.estimate(method=sequential)
print(exit_result.summary(protocol="fingerprint"))

Longitudinal TMLE (2 time points, n = 4000)

parameter                   estimate  std. error  95% CI            p-value
--------------------------  --------  ----------  ----------------  -------
risk_regimen[always @ t=1]  0.1526    0.0086      [0.1357, 0.1695]  <1e-4  
risk_regimen[always @ t=2]  0.3278    0.0135      [0.3013, 0.3544]  <1e-4  
risk_regimen[never @ t=1]   0.2673    0.0112      [0.2453, 0.2894]  <1e-4  
risk_regimen[never @ t=2]   0.4854    0.0180      [0.4500, 0.5207]  <1e-4  

  time points: 2
  outcome: survival, event indicator at plan_exit_p1, plan_exit_p2
  horizons reported: t = 1, 2
  outcome family: binomial
  regimens: always=(1/1), never=(0/0)
  cross-fitting: none -- nuisances fitted in sample
  g_bounds: fixed [0.01, 1] on each cumulative treatment-and-censoring probability (package default; R ltmle-compatible heuristic -- inspect truncation diagnostics)
  confidence level: 95%
  random_state: 41
  causal estimand: mean outcome under each declared regime


**What this output tells you.** The table has one row per plan per horizon.

| plan | risk by 30 days | risk by 60 days | population value at 60 days |
| --- | --- | --- | --- |
| always | 0.1526 | 0.3278 | 0.3172 |
| never | 0.2673 | 0.4854 | 0.4552 |

The summary below the table is long. These lines matter most for this fit.

| summary line | what it says |
| --- | --- |
| `horizons reported: t = 1, 2` | the horizons. Each one is a time point of the fit |
| `regimens: always=(1/1), never=(0/0)` | the arm each plan assigns at the two nodes |
| `cross-fitting: none` | the in-sample construction |
| `g_bounds: fixed [0.01, 1]` | the bound on each cumulative probability of following the plan and staying tracked |
| `clusters = 200` | the navigator teams, with a cluster-robust variance |
| `always: 1007 of 4000 units followed it throughout` | the followers of each plan: 1007 for always and 608 for never. The largest weights are 22.2 and 27.4 |

**A horizon is one of the fit's own time points, not a day.** Here time point 2 ends on day 60.
The summary prints no `reference` line, because a `RegimeMean` fit reports no contrast. Step 7
computes the always-minus-never contrasts from this fit.

The simultaneous bands are built to hold all four parameters at once with 95% probability. Their
critical value is 2.424 against 1.960, so they are wider than the pointwise intervals.

The generator includes a nonlinear second-period hazard. These compact learners are therefore not a
claim that each nuisance regression is correctly specified.

## Step 7: the retention curve and the plan difference

A retention team reads the share still enrolled, not the share that left. The code reads the same
fit on the survival scale.

The code then compares the plans at each horizon with the delta method. That method uses the
influence curves the fit already holds, which the
[inference entry](../technical-reference/inference.md) defines.

In [7]:
survival_curve = exit_result.curve(scale="survival")
columns = ["estimand", "parameter", "time", "psi", "ci_lower", "ci_upper"]
print(survival_curve[columns].round(4).to_string(index=False))
print()


def difference(result, first, second):
    """Delta-method difference of two estimates that one fit already holds."""
    return result.contrast(lambda psi: psi[0] - psi[1], [first, second], name=f"{first} - {second}")


exit_differences = {
    horizon: difference(
        exit_result, f"risk_regimen[always @ t={horizon}]", f"risk_regimen[never @ t={horizon}]"
    )
    for horizon in (1, 2)
}
print("always minus never, cumulative risk of plan exit:")
for horizon, estimate in exit_differences.items():
    low, high = estimate.ci
    population = exit_truth[f"ate_regimen[always vs never @ t={horizon}]"]
    print(
        f"  t={horizon}: {estimate.psi:.4f}  SE={estimate.std_error:.4f}  "
        f"CI=({low:.4f}, {high:.4f})  population {population:.4f}"
    )

                      estimand                  parameter  time    psi  ci_lower  ci_upper
survival_regimen[always @ t=1] risk_regimen[always @ t=1]     1 0.8474    0.8305    0.8643
survival_regimen[always @ t=2] risk_regimen[always @ t=2]     2 0.6722    0.6456    0.6987
 survival_regimen[never @ t=1]  risk_regimen[never @ t=1]     1 0.7327    0.7106    0.7547
 survival_regimen[never @ t=2]  risk_regimen[never @ t=2]     2 0.5146    0.4793    0.5500

always minus never, cumulative risk of plan exit:
  t=1: -0.1147  SE=0.0140  CI=(-0.1423, -0.0872)  population -0.1082
  t=2: -0.1576  SE=0.0227  CI=(-0.2020, -0.1131)  population -0.1380


**What this output tells you.** `curve()` returns one row per plan per horizon, with a `time`
column. On the survival scale a level row reports one minus the risk, $1 - F$, and its interval
bounds swap. For example, the always plan keeps 0.6722 of patients enrolled at 60 days, with the
interval (0.6456, 0.6987). The `estimand` column names the survival quantity, and the `parameter`
column names the risk estimate behind it.

| horizon | always minus never | 95% interval | population value |
| --- | --- | --- | --- |
| 30 days | -0.1147 | (-0.1423, -0.0872) | -0.1082 |
| 60 days | -0.1576 | (-0.2020, -0.1131) | -0.1380 |

On this draw both risk differences are negative, and the 60-day reduction is larger. Under the
always plan, more of the population would remain enrolled at both horizons. The population values
have the same pattern, and each interval contains its population value. That is one draw, not a
coverage result.

`difference()` gives the same estimate and standard error as a separate `RegimeContrast` fit,
without a second backward recursion. The semantic test for this page checks that equality.

## Step 8: two competing clinical events, with disenrollment as censoring

A patient can be readmitted, or can die before readmission. Either event ends the patient's time at
risk for the other. The two causes are mutually exclusive and absorbing.

| cause | what it is | can the program move it? |
| --- | --- | --- |
| readmission | the first unplanned readmission after discharge | possibly. Navigation can change access and adherence |
| death | death before a recorded readmission | possibly. Any claim needs a plausible pathway and the same identification review |

A patient can also leave the plan. The plan then receives no later claims, so it cannot record a
readmission after that date. Disenrollment is therefore censoring, not a competing event. A
readmission can still happen after it.

This step draws a separate synthetic law with its own patients. A mapping of cause to one indicator
column per time point declares competing risks. The `censoring=` role declares enrollment in each
period, and the design reuses `navigation_nodes`. The code copies the protocol and prints only the
fields it replaces.

In [8]:
from cleverly.datasets import make_longitudinal_competing

event_frame, event_truth = make_longitudinal_competing(n=4_000, seed=53, censoring=True)
event_frame = event_frame.rename(
    columns={
        "W1": "age",
        "W2": "baseline_readiness",
        "A1": "navigation_p1",
        "C1": "enrolled_p1",
        "L2": "identified_needs",
        "A2": "navigation_p2",
        "C2": "enrolled_p2",
        "D1": "death_p1",
        "D2": "death_p2",
        "R1": "readmission_p1",
        "R2": "readmission_p2",
    }
)
event_protocol = replace(
    exit_protocol,
    treatment_strategies=(
        "Offer navigation at the start of each 30-day period until readmission or death",
        exit_protocol.treatment_strategies[1],
    ),
    outcome=(
        "First unplanned readmission, and death before a recorded readmission, "
        "recorded at the end of each 30-day period"
    ),
    intercurrent_event_handling=(
        "Treat death before readmission as a competing event that stays in the history",
        "Treat disenrollment before readmission or death as censoring, so the target is the "
        "incidence had every patient stayed enrolled (hypothetical strategy)",
        "Make no navigation decision after either event",
        exit_protocol.intercurrent_event_handling[2],
    ),
    assumption_rationale=(
        "Recorded history covers the measured common causes of navigation, disenrollment, "
        "readmission, and death in each period",
        *exit_protocol.assumption_rationale[1:],
    ),
)
event_study = CausalStudy(
    event_frame,
    design=LongitudinalTreatment(
        outcome={
            "readmission": ("readmission_p1", "readmission_p2"),
            "death": ("death_p1", "death_p2"),
        },
        censoring=("enrolled_p1", "enrolled_p2"),
        **navigation_nodes,
    ),
    protocol=event_protocol,
)
event_effect = event_study.identify(levels_by_plan)
print("rows and columns:", event_frame.shape)
print("disenrolled in period 1:   ", int(event_frame["enrolled_p1"].eq(0).sum()))
print("readmitted in period 1:    ", int(event_frame["readmission_p1"].eq(1).sum()))
print("died in period 1:          ", int(event_frame["death_p1"].eq(1).sum()))
print("at risk entering period 2: ", int(event_frame["navigation_p2"].notna().sum()))
print("disenrolled in period 2:   ", int(event_frame["enrolled_p2"].eq(0).sum()))
print()
print("\n".join(event_effect.summary().splitlines()[:2]))
print()
print("causal study protocol:", event_protocol.fingerprint)
for field in (
    "treatment_strategies",
    "outcome",
    "intercurrent_event_handling",
    "assumption_rationale",
):
    print(f"{field}: {getattr(event_protocol, field)}")
print()
print("known values of the synthetic law, cumulative incidence of death:")
for key in ("cif_regimen[never, death @ t=2]", "cif_regimen[always, death @ t=2]"):
    print(f"  {key:34s} {event_truth[key]:.4f}")

rows and columns: (4000, 11)
disenrolled in period 1:    451
readmitted in period 1:     457
died in period 1:           253
at risk entering period 2:  2839
disenrolled in period 2:    233

mean outcome under each declared regime
identified by explicit-adjustment: sequential g-formula under the declared treatment regimen

causal study protocol: 9b9cb1dda3683cf0
treatment_strategies: ('Offer navigation at the start of each 30-day period until readmission or death', 'Offer no navigation at either decision')
outcome: First unplanned readmission, and death before a recorded readmission, recorded at the end of each 30-day period
intercurrent_event_handling: ('Treat death before readmission as a competing event that stays in the history', 'Treat disenrollment before readmission or death as censoring, so the target is the incidence had every patient stayed enrolled (hypothetical strategy)', 'Make no navigation decision after either event', 'Analyze each navigation offer regardless of complet

**What this output tells you.** The counts show the ways to leave the first risk set.
In period 1, 451 patients left the plan, 457 were readmitted, and 253 died. That leaves 2839 patients at risk
entering period 2. Of these, 233 left the plan in period 2.

The identification is the same sequential g-formula. The protocol now has the fingerprint
`9b9cb1dda3683cf0`. Four fields changed: the treatment strategies, the outcome, the
intercurrent-event handling, and the assumption rationale.

| event after time zero | the protocol's rule |
| --- | --- |
| death before readmission | a competing event that stays in the history |
| disenrollment before readmission or death | censoring. The target is the incidence had every patient stayed enrolled |
| readmission or death | no later navigation decision |

The hypothetical target of continued enrollment needs two more assumptions than Step 5 lists.

| assumption | what it says here | can the data check it? |
| --- | --- | --- |
| independent censoring | given the recorded history, leaving the plan in a period is independent of the later readmission and death risks | no |
| positivity for staying enrolled | every kind of patient on each plan has some chance of staying enrolled through each period | partly, through the support report |

The generator is `make_longitudinal_competing`. It has no cluster option, so this study declares no
navigator teams. In this synthetic law, 60-day death risk is 21% under no navigation and 7% under
navigation at both periods.

Steps 2 and 8 use two separate synthetic laws. The competing law reuses the event hazards of the
plan-exit law, and it then draws which cause occurred. Its readmission and death incidences
therefore add up to the plan-exit risk of Step 2. In a real plan, readmission does not end
enrollment, so the two outcomes differ.

## Step 9: estimate the cause-specific incidences

The fit reuses the method from Step 6. It reports one cumulative incidence per plan, per cause, per
horizon. The first table puts each estimate beside its population value. Its `distance_in_se`
column is the gap between the two, in standard errors. The code then compares the plans for each
cause and sums the causes. Event-free survival is one minus that sum, so the code computes it from
the sum.

In [9]:
from scipy.stats import norm

event_levels = event_effect.estimate(method=sequential)

# The generator names readmission "relapse", so its truth keys use that word.
generator_cause = {"readmission": "relapse", "death": "death"}
level_columns = ["estimand", "psi", "std_err", "ci_lower", "ci_upper"]
level_table = event_levels.to_frame()[level_columns].copy()
level_table["population"] = [
    event_truth[name.replace("readmission", "relapse")] for name in level_table["estimand"]
]
level_table["distance_in_se"] = (level_table["psi"] - level_table["population"]) / level_table[
    "std_err"
]
print(level_table.round(4).round({"distance_in_se": 1}).to_string(index=False))
print()

event_differences = {}
print("always minus never, cause-specific cumulative incidence:")
for cause, source in generator_cause.items():
    for horizon in (1, 2):
        estimate = difference(
            event_levels,
            f"cif_regimen[always, {cause} @ t={horizon}]",
            f"cif_regimen[never, {cause} @ t={horizon}]",
        )
        event_differences[cause, horizon] = estimate
        low, high = estimate.ci
        population = event_truth[f"ate_regimen[always vs never, {source} @ t={horizon}]"]
        distance = (estimate.psi - population) / estimate.std_error
        print(
            f"  {cause:11s} t={horizon}: {estimate.psi:.4f}  SE={estimate.std_error:.4f}  "
            f"CI=({low:.4f}, {high:.4f})  population {population:.4f}  "
            f"distance {distance:.1f} SE"
        )
print()
incidence_totals = event_levels.incidence_total()
print(incidence_totals[["regimen", "time", "total", "std_err", "excess"]].round(4))
print()

# Event-free survival is one minus the sum of the causes, with the same standard error.
critical = norm.ppf(0.975)
event_free = incidence_totals[["regimen", "time"]].assign(
    event_free=1.0 - incidence_totals["total"], std_err=incidence_totals["std_err"]
)
event_free["ci_lower"] = event_free["event_free"] - critical * event_free["std_err"]
event_free["ci_upper"] = event_free["event_free"] + critical * event_free["std_err"]
event_free["population"] = [
    1.0
    - event_truth[f"cif_regimen[{row.regimen}, relapse @ t={row.time}]"]
    - event_truth[f"cif_regimen[{row.regimen}, death @ t={row.time}]"]
    for row in event_free.itertuples()
]
print("event-free survival, one minus the incidence total:")
print(event_free.round(4).to_string(index=False))

                              estimand    psi  std_err  ci_lower  ci_upper  population  distance_in_se
cif_regimen[always, readmission @ t=1] 0.1192   0.0076    0.1042    0.1341      0.1187             0.1
cif_regimen[always, readmission @ t=2] 0.2508   0.0116    0.2281    0.2735      0.2467             0.4
      cif_regimen[always, death @ t=1] 0.0303   0.0042    0.0220    0.0385      0.0310            -0.2
      cif_regimen[always, death @ t=2] 0.0722   0.0070    0.0585    0.0859      0.0705             0.3
 cif_regimen[never, readmission @ t=1] 0.1364   0.0084    0.1200    0.1529      0.1452            -1.0
 cif_regimen[never, readmission @ t=2] 0.2285   0.0141    0.2008    0.2561      0.2443            -1.1
       cif_regimen[never, death @ t=1] 0.1113   0.0076    0.0965    0.1262      0.1127            -0.2
       cif_regimen[never, death @ t=2] 0.2098   0.0140    0.1824    0.2373      0.2109            -0.1

always minus never, cause-specific cumulative incidence:
  readmission t

**What this output tells you.** Read the differences first.

| cause | 30 days | 60 days | population values |
| --- | --- | --- | --- |
| readmission | -0.0172 | 0.0223 | -0.0265 and 0.0024 |
| death | -0.0811 | -0.1376 | -0.0817 and -0.1404 |

On this draw the readmission difference is negative at 30 days and positive at 60 days. The
population readmission difference also crosses zero between the two horizons. The death difference
is negative at both horizons, and the reduction is larger at 60 days.

Each of the 12 intervals contains its population value on this draw. The largest distance is 1.1
standard errors. Over the 800 draws of the review probe in the trust section, the 12 intervals
missed 0.64 times per draw. Twelve intervals that each cover 95% of draws miss 0.60 times per draw
on average.

The page's treatment and disenrollment learners have the form of the law's mechanisms: logistic
models with main terms of the recorded history. The outcome regressions do not match the law,
which multiplies an event hazard by a cause share and has a `tanh` term. TMLE is then consistent
although the outcome regressions are misspecified. The
[competing-risk study](../technical-reference/method-evidence/ordinary-competing-risk-longitudinal-tmle.md)
measures that property with the true mechanism supplied. This page fits a correctly specified
mechanism, which no registered study covers.

Navigation lowers death risk in this law, so more navigated patients remain at risk of readmission.
The readmission difference therefore includes a pathway through death. It is not a negative
control.

The two laws share their event hazards. The 60-day plan-exit difference of Step 2, -0.1380,
therefore equals the death difference, -0.1404, plus the readmission difference, 0.0024. In these
laws, the plan-exit risk of Steps 6 and 7 falls because fewer patients die.

`incidence_total()` sums the causes per plan per horizon. It does not renormalize them, because
each cause solves its own score equation. At 60 days the totals are 0.3230 for always and 0.4383
for never. The `excess` column is how far a total rises above one, and it is 0.0 in every row.

Event-free survival is one minus that **sum**, with the same standard error. One minus one cause's
incidence is not event-free survival. At 60 days, event-free survival is 0.6770 under always and
0.5617 under never. Each event-free interval contains its population value on this draw.

## Step 10: the failure mode, coding death as censoring

The estimates above are total effects in the sense of Young and colleagues (2020). Each readmission
incidence counts every pathway from navigation, including the pathway through death.

A common shortcut codes death as a censoring event. The fit then runs, but it targets a different
estimand. The code builds a new censoring indicator, $K_k = C_k (1 - D_k)$. A patient stays in the
readmission risk set only while enrolled and alive. The code then fits the shortcut and writes the
protocol it implies.

The generator returns no value for the estimand of the shortcut. The code therefore writes that
value as a constant. The review probe of the trust section computes it by quadrature from the
generator's structural equations.

In [10]:
# A patient leaves the readmission risk set at disenrollment or at death, whichever is first:
# the new censoring indicator is K_k = C_k (1 - D_k).
kept_p1 = event_frame["enrolled_p1"] * (1.0 - event_frame["death_p1"].fillna(0.0))
through_p1 = kept_p1.eq(1) & event_frame["readmission_p1"].eq(0)
kept_p2 = (event_frame["enrolled_p2"] * (1.0 - event_frame["death_p2"].fillna(0.0))).where(
    through_p1
)
death_as_censoring = event_frame.assign(
    enrolled_alive_p1=kept_p1,
    enrolled_alive_p2=kept_p2,
    readmission_p1=event_frame["readmission_p1"].where(kept_p1.eq(1)),
    readmission_p2=event_frame["readmission_p2"].where(
        kept_p2.eq(1) | event_frame["readmission_p1"].eq(1)
    ),
)
eliminated_protocol = replace(
    event_protocol,
    outcome="First unplanned readmission, recorded at the end of each 30-day period",
    intercurrent_event_handling=(
        "Treat death as censoring, so the target is readmission risk if death were "
        "prevented (hypothetical strategy)",
        event_protocol.intercurrent_event_handling[1],
        "Make no navigation decision after readmission or death",
        event_protocol.intercurrent_event_handling[3],
    ),
    assumption_rationale=(
        "Recorded history covers the measured common causes of navigation, disenrollment, "
        "readmission, and death in each period, including every common cause of death and "
        "readmission",
        *event_protocol.assumption_rationale[1:],
    ),
)
eliminated = (
    CausalStudy(
        death_as_censoring,
        design=LongitudinalTreatment(
            outcome=("readmission_p1", "readmission_p2"),
            censoring=("enrolled_alive_p1", "enrolled_alive_p2"),
            **navigation_nodes,
        ),
        protocol=eliminated_protocol,
    )
    .identify(levels_by_plan)
    .estimate(method=sequential)
)
eliminated_t2 = difference(eliminated, "risk_regimen[always @ t=2]", "risk_regimen[never @ t=2]")

# The generator returns no value for the death-as-censoring functional. These are its values
# in this law, by quadrature over the generator's structural equations, with death before
# readmission in each period.
functional = {"never": 0.304906, "always": 0.264627}
print("protocol fingerprint, death as a competing event:", event_protocol.fingerprint)
print("protocol fingerprint, death as censoring:        ", eliminated_protocol.fingerprint)
print()
print("readmission risk at t=2      never   always  always - never  SE      population")
rows = (
    (
        "death as a competing event",
        event_levels["cif_regimen[never, readmission @ t=2]"].psi,
        event_levels["cif_regimen[always, readmission @ t=2]"].psi,
        event_differences["readmission", 2],
        event_truth["ate_regimen[always vs never, relapse @ t=2]"],
    ),
    (
        "death as censoring",
        eliminated["risk_regimen[never @ t=2]"].psi,
        eliminated["risk_regimen[always @ t=2]"].psi,
        eliminated_t2,
        functional["always"] - functional["never"],
    ),
)
for label, never, always, contrast, population in rows:
    print(
        f"{label:28s} {never:.4f}  {always:.4f}  {contrast.psi:7.4f}         "
        f"{contrast.std_error:.4f}  {population:7.4f}"
    )
print()
print("death as censoring, 95% CI of the difference:", tuple(round(v, 4) for v in eliminated_t2.ci))

protocol fingerprint, death as a competing event:

 9b9cb1dda3683cf0
protocol fingerprint, death as censoring:         cd5958d44e088c86

readmission risk at t=2      never   always  always - never  SE      population
death as a competing event   0.2285  0.2508   0.0223         0.0181   0.0024
death as censoring           0.2879  0.2698  -0.0180         0.0210  -0.0403

death as censoring, 95% CI of the difference: (-0.0592, 0.0231)


**What this output tells you.** The two analyses carry different protocol fingerprints,
`9b9cb1dda3683cf0` and `cd5958d44e088c86`. The death-as-censoring protocol changes three fields:
the outcome, the intercurrent-event handling, and the assumption rationale. Both designs run, and
only the protocol records that the question changed.

| analysis | estimand | 60-day difference (SE) | population value | what it needs |
| --- | --- | --- | --- | --- |
| death as a competing event | the total effect on the cumulative incidence of readmission | 0.0223 (0.0181) | 0.0024 | the assumptions of Steps 5 and 8 |
| death as censoring | the death-as-censoring g-formula functional, which is a controlled direct effect only under exchangeability for death | -0.0180 (0.0210) | -0.0403 | also exchangeability and positivity for death as an intervened node, and a well-defined intervention that prevents death |

On this draw the censored analysis reports a reduction in readmission, and the competing-event
analysis reports an increase. More patients die under the never plan. Censoring at death treats
those patients as if they stayed at risk, which raises readmission risk most under that plan. Here
the never-plan risk rises from 0.2285 to 0.2879, and the always-plan risk rises from 0.2508 to
0.2698.

The interval of the censored difference, (-0.0592, 0.0231), contains both population values on
this draw. One draw cannot separate the two questions. Over the 800 draws of the review probe, the
censored difference was below the competing-event difference on every draw. Its interval contained
the functional on 742 draws and the total effect on 385 draws.

Do not read -0.0180 as a controlled direct effect. This synthetic law draws one event and then
names its cause, so it has no world in which death is prevented. The functional also depends on an
order that the analyst chooses. With death before readmission in each period, as the recoding
assumes, its value is -0.0403. With readmission first, its value is -0.0093.

No registered study covers this construction. [Young et al. (2020)](https://doi.org/10.1002/sim.8471) define both
estimands and their identifying conditions.

## Step 11: diagnostics, what the fit can check

Assess the retention and competing-risk fits separately, because their risk sets have different
shapes. The code prints the operations that ran on the retention fit, and both support reports.

In [11]:
exit_assessment = exit_result.assess()
event_assessment = event_levels.assess()
ledger = exit_assessment.to_frame()
ran = ledger[ledger["status"].isin(["passed", "completed"])]
print(ran[["surface", "check", "status"]].to_string(index=False))
print("needs attention, retention fit:", tuple(item.name for item in exit_assessment.attention))
print("needs attention, competing fit:", tuple(item.name for item in event_assessment.attention))
print()
exit_support = exit_assessment.report("support").to_frame()
print("retention support:")
print(exit_support.round(4).to_string(index=False))
kish_ratio = (exit_support["effective_n"] / exit_support["n_followed"]).min()
print(f"minimum effective-sample-size ratio: {kish_ratio:.1%}")
print()
print("competing-event support:")
print(event_assessment.report("support").to_frame().round(4).to_string(index=False))

   surface           check    status
validation score_equations    passed
validation         support completed
validation nuisance_models completed
needs attention, retention fit: ()
needs attention, competing fit: ()

retention support:


regimen  horizon  time  n_followed  share_assigned_1  max_weight  effective_n  share_truncated  epsilon  converged
 always        1     1        1781               1.0      8.7247    1628.0127              0.0   0.0003       True
 always        2     1        1781               1.0      8.7247    1628.0127              0.0   0.0075       True
 always        2     2        1007               1.0     22.1663     833.3060              0.0  -0.0058       True
  never        1     1        1756               0.0      6.2591    1646.9331              0.0  -0.0017       True
  never        2     1        1756               0.0      6.2591    1646.9331              0.0   0.0013       True
  never        2     2         608               0.0     27.3928     472.9593              0.0   0.0393       True
minimum effective-sample-size ratio: 77.8%

competing-event support:
regimen       cause  horizon  time  n_followed  share_assigned_1  max_weight  effective_n  share_truncated  epsilon  converged

**What this output tells you.** The first table lists the three operations that ran on
the retention fit. The score-equation check `passed`. The support and nuisance reports are
`completed`. A completed result means the calculation ran. It is not a pass. Neither fit has a row
that needs attention.

The minimum effective-sample-size ratio is 77.8%. That is the never plan at the second node: an
effective sample of about 473 from 608 followers.

| support column | what it shows on this draw |
| --- | --- |
| `horizon` and `time` | the parameter's horizon, and the node whose weights the row describes |
| `n_followed` | the patients whose offers match the plan and who stay observed through that node. The never plan keeps 608 at the second node of the retention fit |
| `max_weight` and `effective_n` | the largest weight, 27.4 for never at the second node, and an effective sample of about 473 |
| `share_truncated` | no row reached the bound on either fit |
| `epsilon` | the targeting coefficient. On the competing fit the largest, -0.0679, is on the never-plan readmission row at the second node |
| `converged` | targeting converged for every row |

The competing-event table adds a `cause` column, which names the cause each row targets. Its
weights include the probability of staying enrolled. The
[longitudinal TMLE tutorial](longitudinal-tmle.ipynb#step-9-diagnostics-what-the-fit-can-check)
explains the support columns and shows the truncation curve.

These reports describe weights and fitted equations. They do not establish sequential
exchangeability, independent censoring, or correct nuisance models.

## Step 12: how much a recorded cause moves the estimate

A sensitivity analysis asks how strong an unmeasured confounder would need to be to change the
conclusion. The [sensitivity guide](../user-guide/results-assessment.md#sensitivity-analysis)
describes the operations. No omitted-variable bound is implemented for a longitudinal fit, so this
step uses a benchmark. The code refits the 60-day difference of Step 7 without one recorded
covariate at a time. Each refit uses the same data, method, plans, and protocol.

In [12]:
full_design = exit_study.design
exit_t2 = exit_differences[2]
benchmarks = []
for dropped in ("age", "baseline_readiness", "identified_needs"):
    reduced = replace(
        full_design,
        baseline=tuple(name for name in full_design.baseline if name != dropped),
        time_varying=tuple(
            tuple(name for name in node if name != dropped) for node in full_design.time_varying
        ),
    )
    refit_levels = (
        CausalStudy(exit_frame, design=reduced, protocol=exit_protocol)
        .identify(levels_by_plan)
        .estimate(method=sequential)
    )
    refit = difference(refit_levels, "risk_regimen[always @ t=2]", "risk_regimen[never @ t=2]")
    move = refit.psi - exit_t2.psi
    benchmarks.append(
        {
            "dropped": dropped,
            "psi": round(refit.psi, 4),
            "move": round(move, 4),
            "move_in_se": round(move / exit_t2.std_error, 2),
        }
    )
benchmark = pd.DataFrame(benchmarks)
print(
    f"full history, always minus never at t=2: psi {exit_t2.psi:.4f}, "
    f"standard error {exit_t2.std_error:.4f}"
)
print(benchmark.to_string(index=False))

full history, always minus never at t=2: psi -0.1576, standard error 0.0227
           dropped     psi   move  move_in_se
               age -0.1148 0.0427        1.89
baseline_readiness -0.1269 0.0307        1.35
  identified_needs -0.1245 0.0331        1.46


**What this output tells you.** The full history gives a 60-day difference of
-0.1576 with a standard error of 0.0227. Each row drops one recorded covariate, refits, and divides
the move by that standard error.

| dropped covariate | estimate | move in standard errors | its role in the law |
| --- | --- | --- | --- |
| `age` | -0.1148 | 1.89 | raises the discharge offer, `identified_needs`, and the exit hazard |
| `baseline_readiness` | -0.1269 | 1.35 | lowers both offers and the exit hazard |
| `identified_needs` | -0.1245 | 1.46 | raises the day-31 offer and the second-period exit hazard |

Read the table as a benchmark, not as a bound. Each move is signed, and all three are positive
here. Each covariate sends the offers toward patients with a higher exit hazard. Without it,
navigation looks less protective, and the estimate moves toward zero.

The full-history difference is about seven standard errors below zero. A move of 1.89 standard
errors changes its size, not its sign. An unrecorded cause that acts like these covariates would
make the true reduction larger than the estimate. A move describes the dropped covariate only. An
unrecorded cause can act through other paths, and its move can be larger or have the other sign.

In the review probe of the trust section, all three moves were positive on 800 of 800 draws.

## How far to trust this

A review probe, `reviews/notebook-review/probes/longitudinal-survival-final/sweep.py`, repeats the
fits of this page on 800 draws of its two laws. It is not a registered study. The script `truth.py`
in the same directory recomputes each population value of this page from the structural equations.

| fit | what the review probe measured over 800 draws |
| --- | --- |
| retention, 60-day difference (Step 7) | the interval covered the population value on 763 draws. The standard error averaged 1.03 of the empirical spread |
| competing risks with disenrollment as censoring (Step 9) | each of the 12 intervals covered its population value on 0.935 to 0.961 of the draws |
| death as censoring (Step 10) | the interval covered the functional on 742 draws (0.927). The standard error averaged 0.94 of the empirical spread |

| layer | establishes | does not establish |
| --- | --- | --- |
| assessment overview | which stored checks need attention, and which operations ran | the detail needed to interpret each retained report |
| retained diagnostics | that targeting converged, and how concentrated the fitted weights are | that the nuisance models are right |
| the covariate benchmark | how far the estimate moves when one recorded covariate leaves the history | a bound, or the strength of an unrecorded cause |
| the review probe | the coverage of this page's intervals over 800 draws of its laws with these learners | a registered result, or the behavior on another law |
| the registered studies | the implementation recovers known truths and matches the R reference under their declared conditions | that your identification assumptions hold on your data |

Two registered studies cover parts of the constructions on this page. The
[ordinary survival-curve study](../technical-reference/method-evidence/ordinary-survival-curve-longitudinal-tmle.md)
compares the retention construction against R `ltmle`. The
[ordinary competing-risk study](../technical-reference/method-evidence/ordinary-competing-risk-longitudinal-tmle.md)
compares the competing-event construction against R `lmtp`.

| cited cell | law | nuisance construction | outcome type | fold layer | role | result |
| --- | --- | --- | --- | --- | --- | --- |
| survival accuracy row `ate_regimen[always vs never @ t=2]` | the plan-exit law of this page without teams, n = 2,000, 1,600 draws | mechanisms supplied exactly; main-term quasibinomial sequential regressions fitted | time to event | in sample | positive | coverage 0.9406, SE ratio 0.9938 |
| survival `interval_calibration/static_t2__correctly_specified` | another law: a finite-support binary survival law, n = 2,000, 9,600 draws | every nuisance fitted by saturated cell means, which are correct | time to event | in sample | positive | coverage 0.9276 to 0.9407, SE ratio 0.9411 to 0.9774 |
| competing accuracy row `ate_regimen[always vs never, death @ t=2]` | another law: a finite-support two-cause law with monotone censoring, n = 4,000, 1,600 draws | mechanisms supplied exactly; constant outcome regressions fitted | time to event, two causes | in sample | positive | coverage 0.9519, SE ratio 0.9833 |
| competing `double_robustness/death_static_t2__mechanism_correct` | the same two-cause law, n = 4,000, 1,200 draws | mechanisms supplied exactly; constant outcome regressions, which are wrong | time to event, two causes | in sample | positive | bias -0.0026 to 0.0022, inside the margin of 0.0080; SE ratio 0.9778 |

The values come from each study's `summary.csv` and `properties.csv`.

The calibration cell measures a horizon-two standard error a few percent below the sampling
spread. That finding holds for its own law at n = 2,000, with saturated fitted nuisances. No
registered study measures the horizon-two calibration of this page's laws and learners.

Each study page lists its limits. The competing-risk study excludes clustering and eliminated
competing events. The survival study validates pointwise inference only, so it does not cover the
simultaneous bands. Its page does not describe clustered data. No registered study covers the
learned mechanisms of this page, or the death-as-censoring fit of Step 10.

The [technical entry](../technical-reference/longitudinal-tmle.md#validation-issues-special-to-this-method)
lists the survival and competing-risk evidence and its limits. Nothing in this list validates the
causal reading. That rests on consistency, sequential exchangeability, independent censoring, and
no interference. These are arguments about the program rather than about the fit.

## Where to go next

This page reported one parameter per plan per horizon, and then one per cause as well. A program
comparing many navigation plans wants a summary instead. That is
[MSM projections](msm-projections.ipynb), and the same projection works over regimens and horizons.

The [examples index](index.md#the-program) lists every tutorial in the program.